// oc-server/dist/Server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSettings = exports.getInitialData = exports.Server = void 0;
class Server {
    initial;
    actions = {};
    constructor(initial) {
        this.initial = initial;
    }
    action(name, action) {
        this.actions[name] = action;
        return this;
    }
    getData() {
        return async ({ action: actionName, params, setEmptyResponse, ...context }, cb) => {
            let res;
            try {
                if (actionName && this.actions[actionName]) {
                    const data = params?.data ?? params;
                    res = await this.actions[actionName](data, context);
                }
                else {
                    res = await this.initial(params, context);
                }
            }
            catch (err) {
                cb(err);
                return;
            }
            cb(null, res);
        };
    }
}
exports.Server = Server;
const getInitialData = () => typeof __$$oc_initialData__ !== 'undefined'
    ? __$$oc_initialData__
    : {};
exports.getInitialData = getInitialData;
const getSettings = () => typeof __$$oc_Settings__ !== 'undefined' ? __$$oc_Settings__ : {};
exports.getSettings = getSettings;
